doc/design/block_execution.md

```diff
@@ -0,0 +1,147 @@
+
+3. Variable Places. Devices often have their onboard memories, and it is usually much more efficient for them to operator variables in the onboard memories. The execution plan should include the placement of variables in a block.
+
+
+Here are the features we want to support:
+
+1. Baseline
+  - a single thread runs a block on a single device, e.g., a CPU or a GPU.
+
+2. Data Parallelism
+  - multi-threads in a stand-alone machine: Only one copy of parameters needs to be storaged in CPU memory.
+  - multi-GPUs in a stand-alone machine: Each GPU card will have a full copy of parameters in its own GPU memory.
+  - multi-threads/multi-GPUs in a cluster: Each node in the cluster supports multi-threads/multi-GPUs.
+
+3. Model Parallelism
+  - Operators in a block can locate in different devices(CPU/GPU/FPGA) in a stand-alone machine. Users have to set device id for every operator manually.
+  - We should notice another situation where operators in a block can be executed in different CPU threads or CUDA streams to achieve high performace, but they are actually in the same device. It's scheduled by Paddle automatically, and users don't need to set specific CPU thread id or CUDA stream id. 
+
+4. Long-term
+  - Paddle schedules CPU/GPU/FPGA in a stand-alone machine automatically, makes full use of them to get best performance when executing a neural network.
+  - Paddle schedules CPU/GPU/FPGA in a cluster automatically, and overlaps communication and computation well to get best performance.
+
+  
+  
+## Analysis
+
+The execution of a neural network topology in PaddlePaddle is divided into two parts, complie-time and run-time.
+
+At complie-time, a ProgramDesc will be generated. At run-time, the ProgramDesc will be executed on specific hardwares.
+
+We can refer to the design of [computation-graphs](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/refactorization.md#computation-graphs).
+
+### Compile-time analysis
+
+In order to realize features mentioned in overview, we have to do several transform pass to ProgramDesc:
+
+- forward pass: generate OpDesc and VarDesc for forward operators
+
+- backward pass: generate OpDesc and VarDesc for backward operators
+
+- placement pass: generate OpDesc and VarDesc for send/recv/copy operators. Data Parallelism and Model Parallelism will be considered at this stage. We have a placement policy to set concrete device id for every operator. And send/recv/copy operators will be inserted at correct position. 
+
+- optimize pass: generate OpDesc and VarDesc for optimize operators
+
+Each pass will modify the global ProgramDesc.
+
+### Run-time analysis
```

Comment:
Maybe mention "Converter" in section unified-pass-interface is the module that handle the analysis.

Reply:
Done